### ISSUE TITLE
Poor performance of basic operations on CPU

### ISSUE THREAD

**Issue by FloopCZ** [arrayfire/arrayfire#1449]
Although arrayfire/arrayfire#1440 increased the performance of my ArrayFire code about 6 times, it is still about 3 times slower than its NumPy version (written in Python). Consider the following simple operations:
```c++
B = af::sum(A, 0);
B = af::sqrt(af::sum(A, 0));
B = af::sqrt(af::sum(A * A, 0)); // column-wise L2 norm
```
Their direct NumPy counterparts are:
```python
B = np.sum(A, 0);
B = np.sqrt(np.sum(A, 0));
B = np.sqrt(np.sum(A * A, 0));
```
Let us measure the performance in both ArrayFire and NumPy by repeating the operations 10000 times on matrices of size 512x512.
Arrayfire:
```
ArrayFire v3.4.0 (CPU, 64-bit Linux, build 745e47c)
[0] Unknown: Unknown, 15880 MB, Max threads(1) 
arrayfire - sum(A, 0): 6.41246
arrayfire - sqrt(sum(A, 0)): 6.46102
arrayfire - sqrt(sum(A * A, 0)) (L2 norm): 18.1683
arrayfire - matmul: 6.66702
```
Numpy:
```
numpy - sum(A, 0): 0.526026725769043
numpy - sqrt(sum(A, 0))): 0.5332002639770508
numpy - sqrt(sum(A * A, 0)) (L2 norm): 0.9607725143432617
```
As the results suggest, the performance is much worse with ArrayFire. Additionally, it is kind of surprising that elementwise operations (O(N^2)) take longer than matrix multiplication (O(N^3)) for matrices of size 512x512.

I was not yet able to identify the source of this inefficiency, you might have a better insight.

Full code [C++](https://gist.github.com/FloopCZ/79d14dc1bf8b97a1468faf0c1a32cb8c), [Python](https://gist.github.com/FloopCZ/12f125d2fb5f7a7bb38a79104473f90f)

Note: Both ArrayFire and NumPy are linked against Intel MKL.

**Comment by pavanky** [comment]
This is a [known issue](https://github.com/arrayfire/arrayfire/issues/1226). We did not target CPU backend for performance. It was a fallback option if CUDA and OpenCL were not available on the machine. OpenCL on CPU gives the desired performance in most cases.

That said we are addressing some of that for 3.4.0. 

Closing because it is a duplicate.

**Comment by FloopCZ** [comment]
Ok, that sounds reasonable. Unfortunately, for me, the OpenCL backend has similar performance on the simple operations, plus the matrix multiplication is incredibly slow:
```
Device[1] has no support for OpenGL Interoperation
ArrayFire v3.3.2 (OpenCL, 64-bit Linux, build f65dd97)
-0- NVIDIA  : Quadro 2000M, 2047 MB
[1] INTEL   : Intel(R) Core(TM) i7-2820QM CPU @ 2.30GHz, 16001 MB
arrayfire - sum(A, 0): 3.608
arrayfire - sqrt(sum(A, 0)): 3.8374
arrayfire - sqrt(sum(A * A, 0)) (L2 norm): 5.23145
arrayfire - matmul: 119.737
```
Using htop, I can see that the CPU is running on full power on all 8 virtual threads, however, the real time is very slow anyway.

Using OpenCL on GPU, the results are slightly better, but still slow (at least compared to NumPy on CPU):
```
Device[1] has no support for OpenGL Interoperation
ArrayFire v3.3.2 (OpenCL, 64-bit Linux, build f65dd97)
[0] NVIDIA  : Quadro 2000M, 2047 MB
-1- INTEL   : Intel(R) Core(TM) i7-2820QM CPU @ 2.30GHz, 16001 MB
Device count: 2
arrayfire - sum(A, 0): 2.79402
arrayfire - sqrt(sum(A, 0)): 2.83406
arrayfire - sqrt(sum(A * A, 0)) (L2 norm): 3.70326
arrayfire - matmul: 18.7813
```
CUDA backend is the fastest of all three, but still cca 2x slower than the NumPy.
```
ArrayFire v3.3.2 (CUDA, 64-bit Linux, build f65dd97)
Platform: CUDA Toolkit 7.5, Driver: 352.93
[0] Quadro 2000M, 2048 MB, CUDA Compute 2.1
Device count: 1
arrayfire - sum(A, 0): 1.74765
arrayfire - sqrt(sum(A, 0)): 1.83636
arrayfire - sqrt(sum(A * A, 0)) (L2 norm): 2.6903
arrayfire - matmul: 14.3694
```
Am I doing something wrong?

**Comment by pavanky** [comment]
1) you are getting 2.8s for 10000 iterations. That is about 0.28ms per iteration. There's overhead of launching kernels and other things associated when using OpenCL. At smaller sizes like 512 x 512, the times you are seeing have large kernel launch overheads.

2) There is also an issue with OpenCL backend compiling the kernels on the fly. Can you try running the tests *once* (including evals) before timing it to notice the difference ?

3) OpenCL matrix multiply is optimized for GPUs (AMD GPUs in particular). To enable the user to get the best out the hardware, we enabled the following environment variable: `AF_OPENCL_CPU_OFFLOAD`. If you export that environment variable with value `1`, arrayfire uses MKl (if you used our installer) or another blas (if you built from source) available on the system. This behavior is disabled by default because it is technically not using OpenCL.

4) I tried to benchmark the code my machine with these changes. It looks like numpy is faster than CPU + OpenCL even after this. But something I noticed was that numpy's `np.sum(A, d)` would be equivalent of `af::sum(A, A.dims() - d)` because of the way memory is laid out. When I changed the arrayfire code to use `af::sum(A, 1)` in arrayfire, here are some results (I ended up using 1000 iterations instead of 10000).

```
AF_OPENCL_CPU_OFFLOAD=1 AF_OPENCL_DEFAULT_DEVICE=2 ./a.out 
ArrayFire v3.4.0 (OpenCL, 64-bit Linux, build b09bee6)
-0- AMD     : Hawaii, 7368 MB
-1- NVIDIA  : GeForce GTX 950, 2047 MB
[2] INTEL   : Intel(R) Core(TM) i7-3770K CPU @ 3.50GHz, 14997 MB
-3- AMD     : Intel(R) Core(TM) i7-3770K CPU @ 3.50GHz, 14997 MB
Running size: 512
Before caching
arrayfire - sum(A, 1): 0.115503
arrayfire - sqrt(sum(A, 1)): 0.0916993
arrayfire - sqrt(sum(A * A, 1)) (L2 norm): 0.191266
arrayfire - matmul: 2.80714
After caching
arrayfire - sum(A, 1): 0.107088
arrayfire - sqrt(sum(A, 1)): 0.0743911
arrayfire - sqrt(sum(A * A, 1)) (L2 norm): 0.166777
arrayfire - matmul: 3.1563
Running size: 1024
Before caching
arrayfire - sum(A, 1): 0.239752
arrayfire - sqrt(sum(A, 1)): 0.211787
arrayfire - sqrt(sum(A * A, 1)) (L2 norm): 0.533824
arrayfire - matmul: 21.5394
After caching
arrayfire - sum(A, 1): 0.253705
arrayfire - sqrt(sum(A, 1)): 0.239988
arrayfire - sqrt(sum(A * A, 1)) (L2 norm): 0.55973
arrayfire - matmul: 21.7654
```

The equivalent numpy results
```
$ python bench_norm.py 
Size: 512
numpy - sum(A, 0): 0.09518694877624512
numpy - sqrt(sum(A, 0))): 0.07397651672363281
numpy - sqrt(sum(A * A, 0)) (L2 norm): 0.12102055549621582
Size: 1024
numpy - sum(A, 0): 0.18867158889770508
numpy - sqrt(sum(A, 0))): 0.1903548240661621
numpy - sqrt(sum(A * A, 0)) (L2 norm): 0.5454561710357666
```

As you can see these numbers are more or less identical.

When I keep arrayfire the same, but change numpy, here are the results:
```
AF_OPENCL_CPU_OFFLOAD=1 AF_OPENCL_DEFAULT_DEVICE=2 ./a.out 
ArrayFire v3.4.0 (OpenCL, 64-bit Linux, build b09bee6)
-0- AMD     : Hawaii, 7358 MB
-1- NVIDIA  : GeForce GTX 950, 2047 MB
[2] INTEL   : Intel(R) Core(TM) i7-3770K CPU @ 3.50GHz, 14997 MB
-3- AMD     : Intel(R) Core(TM) i7-3770K CPU @ 3.50GHz, 14997 MB
Running size: 512
Before caching
arrayfire - sum(A, 0): 0.331304
arrayfire - sqrt(sum(A, 0)): 0.281203
arrayfire - sqrt(sum(A * A, 0)) (L2 norm): 0.372971
arrayfire - matmul: 2.03145
After caching
arrayfire - sum(A, 0): 0.292155
arrayfire - sqrt(sum(A, 0)): 0.260863
arrayfire - sqrt(sum(A * A, 0)) (L2 norm): 0.351649
arrayfire - matmul: 2.85336
Running size: 1024
Before caching
arrayfire - sum(A, 0): 0.58577
arrayfire - sqrt(sum(A, 0)): 0.558537
arrayfire - sqrt(sum(A * A, 0)) (L2 norm): 0.885808
arrayfire - matmul: 14.746
After caching
arrayfire - sum(A, 0): 0.580405
arrayfire - sqrt(sum(A, 0)): 0.557026
arrayfire - sqrt(sum(A * A, 0)) (L2 norm): 0.885418
arrayfire - matmul: 15.4256
```
numpy:
```
$ python bench_norm.py 
Size: 512
numpy - sum(A, 1): 0.09227991104125977
numpy - sqrt(sum(A, 1))): 0.09262967109680176
numpy - sqrt(sum(A * A, 1)) (L2 norm): 0.1529712677001953
Size: 1024
numpy - sum(A, 1): 0.3387112617492676
numpy - sqrt(sum(A, 1))): 0.34192967414855957
numpy - sqrt(sum(A * A, 1)) (L2 norm): 0.6635868549346924
```

As you can see numpy has slowed down, but is still faster the arrayfire. This is because arrayfire has been optimized for cases which have a larger amount of parallelism in them. You can verify this by adding more columns to arrayfire and more rows to numpy.

If you really care about this particular size only, you can open a feature request and we can look into it if others are interested in it as well.

**Comment by pavanky** [comment]
I tried to benchmark the code my machine with these changes. It looks like numpy is faster than CPU + OpenCL even after this. But something I noticed was that numpy's `np.sum(A, d)` would be equivalent of `af::sum(A, A.dims() - d)` because of the way memory is laid out. When I changed the arrayfire code to use `af::sum(A, 1)` in arrayfire, here are some results (I ended up using 1000 iterations instead of 10000).

```
AF_OPENCL_CPU_OFFLOAD=1 AF_OPENCL_DEFAULT_DEVICE=2 ./a.out 
ArrayFire v3.4.0 (OpenCL, 64-bit Linux, build b09bee6)
-0- AMD     : Hawaii, 7368 MB
-1- NVIDIA  : GeForce GTX 950, 2047 MB
[2] INTEL   : Intel(R) Core(TM) i7-3770K CPU @ 3.50GHz, 14997 MB
-3- AMD     : Intel(R) Core(TM) i7-3770K CPU @ 3.50GHz, 14997 MB
Running size: 512
Before caching
arrayfire - sum(A, 1): 0.115503
arrayfire - sqrt(sum(A, 1)): 0.0916993
arrayfire - sqrt(sum(A * A, 1)) (L2 norm): 0.191266
arrayfire - matmul: 2.80714
After caching
arrayfire - sum(A, 1): 0.107088
arrayfire - sqrt(sum(A, 1)): 0.0743911
arrayfire - sqrt(sum(A * A, 1)) (L2 norm): 0.166777
arrayfire - matmul: 3.1563
Running size: 1024
Before caching
arrayfire - sum(A, 1): 0.239752
arrayfire - sqrt(sum(A, 1)): 0.211787
arrayfire - sqrt(sum(A * A, 1)) (L2 norm): 0.533824
arrayfire - matmul: 21.5394
After caching
arrayfire - sum(A, 1): 0.253705
arrayfire - sqrt(sum(A, 1)): 0.239988
arrayfire - sqrt(sum(A * A, 1)) (L2 norm): 0.55973
arrayfire - matmul: 21.7654
```

The equivalent numpy results
```
$ python bench_norm.py 
Size: 512
numpy - sum(A, 0): 0.09518694877624512
numpy - sqrt(sum(A, 0))): 0.07397651672363281
numpy - sqrt(sum(A * A, 0)) (L2 norm): 0.12102055549621582
Size: 1024
numpy - sum(A, 0): 0.18867158889770508
numpy - sqrt(sum(A, 0))): 0.1903548240661621
numpy - sqrt(sum(A * A, 0)) (L2 norm): 0.5454561710357666
```

As you can see these numbers are more or less identical.

When I keep arrayfire the same, but change numpy, here are the results:
```
AF_OPENCL_CPU_OFFLOAD=1 AF_OPENCL_DEFAULT_DEVICE=2 ./a.out 
ArrayFire v3.4.0 (OpenCL, 64-bit Linux, build b09bee6)
-0- AMD     : Hawaii, 7358 MB
-1- NVIDIA  : GeForce GTX 950, 2047 MB
[2] INTEL   : Intel(R) Core(TM) i7-3770K CPU @ 3.50GHz, 14997 MB
-3- AMD     : Intel(R) Core(TM) i7-3770K CPU @ 3.50GHz, 14997 MB
Running size: 512
Before caching
arrayfire - sum(A, 0): 0.331304
arrayfire - sqrt(sum(A, 0)): 0.281203
arrayfire - sqrt(sum(A * A, 0)) (L2 norm): 0.372971
arrayfire - matmul: 2.03145
After caching
arrayfire - sum(A, 0): 0.292155
arrayfire - sqrt(sum(A, 0)): 0.260863
arrayfire - sqrt(sum(A * A, 0)) (L2 norm): 0.351649
arrayfire - matmul: 2.85336
Running size: 1024
Before caching
arrayfire - sum(A, 0): 0.58577
arrayfire - sqrt(sum(A, 0)): 0.558537
arrayfire - sqrt(sum(A * A, 0)) (L2 norm): 0.885808
arrayfire - matmul: 14.746
After caching
arrayfire - sum(A, 0): 0.580405
arrayfire - sqrt(sum(A, 0)): 0.557026
arrayfire - sqrt(sum(A * A, 0)) (L2 norm): 0.885418
arrayfire - matmul: 15.4256
```
numpy:
```
$ python bench_norm.py 
Size: 512
numpy - sum(A, 1): 0.09227991104125977
numpy - sqrt(sum(A, 1))): 0.09262967109680176
numpy - sqrt(sum(A * A, 1)) (L2 norm): 0.1529712677001953
Size: 1024
numpy - sum(A, 1): 0.3387112617492676
numpy - sqrt(sum(A, 1))): 0.34192967414855957
numpy - sqrt(sum(A * A, 1)) (L2 norm): 0.6635868549346924
```

As you can see numpy has slowed down, but is still faster the arrayfire. This is because arrayfire has been optimized for cases which have a larger amount of parallelism in them. You can verify this by adding more columns to arrayfire and more rows to numpy.

If you really care about this particular size only, you can open a feature request and we can look into it if others are interested in it as well.

**Comment by FloopCZ** [comment]
Thank you for such a thorough answer!

1) The proper coalescing (summing rows instead of columns) indeed significantly improved the performance of my code. Thanks for reminding me of that!

2) AF_OPENCL_CPU_OFFLOAD=1 improved the performance of matrix multiplication on CPU. In my opinion, it should be turned on by default when the device is CPU. One does not expect matrix multiplication to be 20x slower.

3) OpenCL really seems to be optimized for the GPU, as on the CPU it is always slower than the NumPy, no matter the size of the matrix. Just a thought: ArrayFire's OpenCL on CPU seems to use all 8 virtual threads on my hyper-threaded CPU. That might be suboptimal - even Intel MKL uses only as many threads as there are physical cores. If you force MKL to use more threads, it drastically slows down. Is there an option to set the number of workers that ArrayFire's OpenCL is allowed to use?

4) On both my computers, the OpenCL backend produces drastically different (and wrong) results than CPU and CUDA backends (on both CPU and GPU). But that is for another issue report.

5) Feature request is a good idea, I usually work with relatively small matrices (upto 512x512).

**Comment by pavanky** [comment]
> OpenCL really seems to be optimized for the GPU, as on the CPU it is always slower than the NumPy, no matter the size of the matrix. 

 Our opinion is the OpenCL backend gets you most of the way there until we improve the CPU backend.

> Is there an option to set the number of workers that ArrayFire's OpenCL is allowed to use?

Intel's OpenCL implementation is opaque to us. AFAIK, there's no way to control this.

> Feature request is a good idea, I usually work with relatively small matrices (upto 512x512).

ArrayFire really thrives when you are using multiple matrices / arrays simultaneously. Are you always working with one small matrix at a time ?

**Comment by UniqueFool** [comment]
that is correct, OpenCL merely gives you functional compatibility, but doesn't provide performance-compatibility - which makes sense actually, kernels will typically need to be customized for different back-ends, as they tend to contain lots of backend/hardware-specific assumptions depending on the platform they're written for 

OpenCL is basically a problem-specification thing, and if the problem is specified in a way that isn't suitable for the back-end, the problem may simply be ill-specified - besides, there's a certain "overhead" that comes with using OpenCL, so that numpy may get away with certain things that would otherwise be considered slow.

**Comment by mchandra** [comment]
4) On both my computers, the OpenCL backend produces drastically different (and wrong) results than CPU and CUDA backends (on both CPU and GPU). But that is for another issue report.

^That is worrisome.

**Comment by UniqueFool** [comment]
http://arxiv.org/pdf/1503.06532.pdf
> OpenCL is usually considered more suitable for runn
ing hundreds or more of short kernels. OpenMP is 
traditionally used to run longer-running code segme
nts in parallel, which would lessen relative thread
management  costs.  [...] OpenCL does invo
lve more programming overhead. The  steeper 
learning  curve  and the  need  to  perform more setup  a
re  an inherent  part  of  working  with  a system 
designed  primarily  for GPU  programming [...] OpenCL and OpenMP’s CPU 
performances are close, making OpenCL a viable 
alternative to OpenMP. When there is a need to perf
orm a large number of thread invocations, OpenCL 
can outperform OpenMP. However, if maximum performa
nce is required, we recommend trying both 
OpenCL and OpenMP to determine which one performs b
etter for the specific task at hand.

**Comment by pavanky** [comment]
> 4) On both my computers, the OpenCL backend produces drastically different (and wrong) results than CPU and CUDA backends (on both CPU and GPU). But that is for another issue report.

> ^That is worrisome.

I'd rather wait until there is a bug report.
